src/base/static/components/organisms/map-measurement-overlay.tsx

```diff
@@ -0,0 +1,373 @@
+  new WebMercatorViewport({
+    ...viewport,
+    width,
+    height,
+  });
+
+const buildMeasurementFeatureCollection = (selectedTool, positions) => {
+  let featureFn;
+  let newPositions;
+  let numPositions = 0;
+  switch (selectedTool) {
+    case "create-polygon":
+      featureFn = polygon;
+      // Ensure the last position of the Polygon matches the first.
+      newPositions = [positions.concat([positions[0]])];
+      numPositions = newPositions[0].length;
+      break;
+    case "create-polyline":
+      featureFn = lineString;
+      newPositions = positions;
+      numPositions = newPositions.length;
+      break;
+    default:
+      featureFn = () => [];
+      newPositions = positions;
+      // eslint-disable-next-line no-console
+      console.error(`Measurement overlay: unsupported tool ${selectedTool}`);
+      break;
+  }
+
+  return featureCollection(
+    positions.map(position => point(position)).concat(
+      // Only add the measurement feature if we have enough positions on
+      // the map to support the given feature type.
+      numPositions >= MIN_POSITIONS[selectedTool]
+        ? featureFn(newPositions)
+        : [],
+    ),
+  );
+};
+
+const redraw = ({
+  project,
+  isDragging,
+  ctx,
+  measurementFeatureCollection,
+  isWithFill,
+  width,
+  height,
+}) => {
+  function projectPoint(lon, lat) {
+    const point = project([lon, lat]);
+    /* eslint-disable-next-line no-invalid-this */
+    this.stream.point(point[0], point[1]);
+  }
+
+  const transform = geoTransform({ point: projectPoint });
+  const path = geoPath()
+    .projection(transform)
+    .context(ctx);
+
+  ctx.clearRect(0, 0, width, height);
+
+  const { features } = measurementFeatureCollection;
+  if (!features) {
+    return;
+  }
+
+  for (const feature of features) {
+    ctx.beginPath();
+    ctx.strokeStyle = "rgba(245, 241, 17, 0.8)"; // Bright yellow.
+    ctx.lineWidth = "2";
+    ctx.fillStyle = "rgba(188, 0, 0, 0.5)"; // Bright red.
+    const geometry = feature.geometry;
+    path({
+      type: geometry.type,
+      coordinates: geometry.coordinates,
+    });
+    isWithFill && ctx.fill();
+    ctx.stroke();
+  }
+};
+
+const formatMeasurement = measurement =>
+  isNaN(measurement)
+    ? null
+    : measurementFormatter.format(measurement.toFixed(1));
+
+const MapMeasurementOverlay = props => {
+  // NOTE: The structure of this featureCollection is such that all but the
+  // last feature are Points representing clicked-on locations, and the final
+  // feature is either a LineString or a Polygon built from those points. The
+  // final feature is used for measurement purposes.
+  const [
+    measurementFeatureCollection,
+    setMeasurementFeatureCollection,
+  ] = React.useState(featureCollection([]));
+  const selectedTool = React.useRef(null);
+  const [measurement, setMeasurement] = React.useState(null);
+  const [units, setUnits] = React.useState(null);
+  const positions = React.useRef([]);
```

Comment:
Not sure if this is desirable, but I made extensive use of `useRef` in this component, due to the fact that the click listener kept having problems with stale state when it ran. I *think* this is because the click listener needs to be registered outside React (using `overlayRef.current._canvas.addEventListener("click", handleOverlayClick);`), and when it's registered it encloses initial state.

I got the idea of using `useRef` to get around this problem from here: https://stackoverflow.com/questions/53845595/wrong-react-hooks-behaviour-with-event-listener